node/core/runtime-api/src/lib.rs

```diff
@@ -68,34 +84,73 @@ impl<Client, Context> Subsystem<Context> for RuntimeApiSubsystem<Client> where
 	}
 }
 
+impl<Client> RuntimeApiSubsystem<Client> where
+	Client: ProvideRuntimeApi<Block> + Send + 'static + Sync,
+	Client::Api: ParachainHost<Block>,
+{
+	/// Spawn a runtime api request.
+	///
+	/// If there are already [`MAX_PARALLEL_REQUESTS`] requests being executed, the request will be buffered.
+	fn spawn_request(&mut self, relay_parent: Hash, request: Request) {
+		let client = self.client.clone();
+		let metrics = self.metrics.clone();
+		let (sender, receiver) = oneshot::channel();
+
+		let request = async move {
+			make_runtime_api_request(
+				client,
+				metrics,
+				relay_parent,
+				request,
+			);
+			let _ = sender.send(());
+		}.boxed();
+
+		if self.active_requests.len() >= MAX_PARALLEL_REQUESTS {
+			self.waiting_requests.push_back((request, receiver));
+		} else {
+			self.spawn_handle.spawn_blocking(API_REQUEST_TASK_NAME, request);
+			self.active_requests.push(receiver);
+		}
+	}
+
+	/// Poll the active runtime api requests.
+	async fn poll_requests(&mut self) {
+		// If there are no active requests, this future should be pending forever.
+		if self.active_requests.len() == 0 {
+			futures::pending!()
+		} if self.active_requests.next().await.is_none() {
+			return
+		}
+
+		if let Some((req, recv)) = self.waiting_requests.pop_front() {
+			self.spawn_handle.spawn_blocking(API_REQUEST_TASK_NAME, req);
+			self.active_requests.push(recv);
+		}
+	}
+}
+
 #[tracing::instrument(skip(ctx, subsystem), fields(subsystem = LOG_TARGET))]
 async fn run<Client>(
 	mut ctx: impl SubsystemContext<Message = RuntimeApiMessage>,
-	subsystem: RuntimeApiSubsystem<Client>,
+	mut subsystem: RuntimeApiSubsystem<Client>,
 ) -> SubsystemResult<()> where
 	Client: ProvideRuntimeApi<Block> + Send + Sync + 'static,
 	Client::Api: ParachainHost<Block>,
 {
 	loop {
-		match ctx.recv().await? {
-			FromOverseer::Signal(OverseerSignal::Conclude) => return Ok(()),
-			FromOverseer::Signal(OverseerSignal::ActiveLeaves(_)) => {},
-			FromOverseer::Signal(OverseerSignal::BlockFinalized(_)) => {},
-			FromOverseer::Communication { msg } => match msg {
-				RuntimeApiMessage::Request(relay_parent, request) => {
-					let client = subsystem.client.clone();
-					let metrics = subsystem.metrics.clone();
-
-					subsystem.spawn_handle.spawn_blocking("polkadot-runtime-api-request", async move {
-						make_runtime_api_request(
-							client,
-							metrics,
-							relay_parent,
-							request,
-						)
-					}.boxed())
-				},
-			}
+		select! {
+			req = ctx.recv().fuse() => match req? {
+				FromOverseer::Signal(OverseerSignal::Conclude) => return Ok(()),
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(_)) => {},
+				FromOverseer::Signal(OverseerSignal::BlockFinalized(_)) => {},
+				FromOverseer::Communication { msg } => match msg {
+					RuntimeApiMessage::Request(relay_parent, request) => {
+						subsystem.spawn_request(relay_parent, request);
+					},
+				}
+			},
+			_ = subsystem.poll_requests().fuse() => {},
```

Comment:
Here it seems that in case there were no message from the overseer we will end up in `poll_requests`. In case there is a next `active_requests` we will await on it and as soon as the request finishes we return control here.

During this time span, we don't serve the `ctx.recv()`? So for example, if a conclude signal comes we will wait until the requests finishes. Is that right? 

Not that bad I guess, just testing my understanding.